### PR TITLE
Refactor MaxCosts and try max bridging

### PR DESCRIPTION
**Pull request by mykelp**

### CODE REVIEW

**Review by hayesgm — APPROVED** [review]
I need to dig in further, but I generally like it. It's big enough that we'd need to sit down to discuss in full, but if it passes acceptance tests and we can hook into integration tests, I think we move forward.

File: src/builder/QuarkBuilderBase.sol
Comment (kevincheng96) [inline src/builder/QuarkBuilderBase.sol:1]: `QUARK_BUILDER_VERSION` should be updated

**Review by kevincheng96 — APPROVED** [review]
I think the logic generally looks good. 

Is this hard blocked by the stack too deep issues? Changing the optimizer settings means the script addresses will all be different. Not sure if that's an issue though

File: src/builder/QuarkBuilderBase.sol
[the remark this answers is not in the document]
Comment (mykelp) [reply src/builder/QuarkBuilderBase.sol:1]: Yeah, I think so. Another option is to leave it for actions that don't need to support max, but probably cleaner to remove it

**Comment by hayesgm** [comment]
> I think the logic generally looks good.
> 
> Is this hard blocked by the stack too deep issues? Changing the optimizer settings means the script addresses will all be different. Not sure if that's an issue though

We can always revert back once we handle stack too deep, but overall, there's no strict problem with having Paycall address change aside of the fact we'll end up having to "re-deploy" it.

File: src/builder/QuarkBuilderBase.sol
Comment (mykelp) [reply src/builder/QuarkBuilderBase.sol:1]: added

File: src/builder/QuarkBuilderBase.sol [hunk 1]
```diff
@@ -281,27 +266,107 @@ contract QuarkBuilderBase {
...
                 actions: actionsArray,
-                payment: payment,
-                useQuotecall: false
+                payment: payment
             });
         }
     }
 
+    /**
+     * @dev Collects assets for an action by checking and bridging asset if necessary to accomodate the intended action.
```
[the remark this answers is not in the document]
Comment (mykelp) [reply src/builder/QuarkBuilderBase.sol]: updated

**Pull request merged**